File: Algorithms/String_Algorithms/LPS/longest_palindromic_substring.rb

puts "Enter the String:"
input = gets.chomp
puts "Entered String is: '#{input}'."

n = input.length

table = Array.new(n) { Array.new(n, false) }

max_len = 1
i = 0
while (i < n)
  table[i][i] = true
  i = i + 1
end

start = 0
i = 0
while i < n - 1
  if (input[i] == input[i + 1])
    table[i][i + 1] = true
    start = i 
    max_len = 2
  end
  i = i + 1
end

k = 3
while k <= n
  i = 0
  while i < (n - k + 1)
    j = i + k - 1
    if (table[i + 1][j - 1] and input[i] == input[j])
      table[i][j] = true

      if k > max_len
        start = i 
        max_len = k
      end
    end
    i = i + 1
  end
  k = k + 1
end

puts "Longest palindromic substring is #{input[start..start+max_len-1]}"

File: Algorithms/String_Algorithms/LPS/longest_palindromic_substring.js
const readline = require('readline');

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

rl.question(
  'Input a string to find the longest palindromic substring: ',
  (str) => {
    console.log(
      `The longest palindromic substring is: ${longestPalSubstr(str)}`
    );
    rl.close();
  }
);

// Examples
// HeyrailataLiarLiveonoevilBye  ==>  railataliar
// Ratsliveonnoevilstar  ==>  ratsliveonnoevilstar
// abraCadabra ==>  aca

function longestPalSubstr(str) {
  const n = str.length; // length of input string

  const s = str.toLowerCase(); // the new string in lowercase to work with

  // Set a matrix to check substrings (default values to false)
  // - matrix[i][j] will be false if substring str[i..j] is not palindrome
  // - otherwise, matrix[i][j] will be true
  const matrix = new Array(n);
  for (let i = 0; i < n; i++) {
    matrix[i] = Array(n).fill(false);
  }

  // All substrings of length 1 are palindromes
  let maxLength = 1;
  for (let i = 0; i < n; i++) {
    matrix[i][i] = true;
  }

  // Check for substring of length 2.
  let start = 0;
  for (let i = 0; i < n - 1; i++) {
    if (s[i] === s[i + 1]) {
      matrix[i][i + 1] = true;
      start = i;
      maxLength = 2;
    }
  }

  // Check for lengths greater than 2.
  // k is length of substring
  for (let k = 3; k <= n; k++) {
    // Starting index
    for (let i = 0; i < n - k + 1; i++) {
      // Ending index of substring from starting index i and length k
      const j = i + k - 1;

      // Checking for substring from ith index to jth index
      // if str.charAt(i+1) to str.charAt(j-1) is a palindrome
      if (matrix[i + 1][j - 1] && s[i] === s[j]) {
        matrix[i][j] = true;
        if (k > maxLength) {
          start = i;
          maxLength = k;
        }
      }
    }
  }
  return s.slice(start, start + maxLength);
}
